File: solutions/clojure/clock/1/src/clock.clj
(ns clock)

(defn clock->string [[hour minute]]
  (format "%02d:%02d" hour minute))

(defn clock [hours minutes]
  (let [all-minutes (+ minutes (* hours 60))
        positive-hours (mod (int (Math/floor (/ all-minutes 60))) 24)
        positive-minutes (mod all-minutes 60)]
    (list positive-hours positive-minutes)))

(defn add-time [[hour minute] time]
  (clock hour (+ minute time)))

File: solutions/haskell/clock/1/src/Clock.hs
module Clock (addDelta, fromHourMin, toString) where

data Clock =
  Time Int Int
  deriving Eq

padWithZeroes2d :: Show a => Num a => a -> [Char]
padWithZeroes2d num =
  if length (show num) == 1 
  then '0' : (show num)
  else show num

fromHourMin :: Int -> Int -> Clock
fromHourMin h m =
  let
    minute = m `mod` 60
    hour = (h + (m `div` 60)) `mod` 24
  in
    Time hour minute

toString :: Clock -> String
toString (Time hour minute) =
  let 
    stringHour = padWithZeroes2d hour
    stringMinute = padWithZeroes2d minute
  in
    stringHour ++ ":" ++ stringMinute 

addDelta :: Int -> Int -> Clock -> Clock
addDelta hour minute (Time h m) = fromHourMin (hour + h) (minute + m) 
